### ISSUE TITLE
Split runtime and compile code into different maven artifacts

### ISSUE THREAD

**Issue by bendem** [casid/jte#2]
From what I understand, the jte to java compiler is included when using this as a maven dependency. Please provide a runtime artifact that does not contain any of this (i.e. jte-compiler and jte-rt).

That way, the maven plugin depends on jte-compiler and my project only has to include jte-rt to work.

Advantage: smaller deployment footprint, less classes/methods included in the IDE's autocomplete, better separation of concern.

Thanks for your time writing this, this looks pretty neat (came here from rocker because of the intellij plugin).

**Comment by casid** [comment]
Hi @bendem!

thanks for the heads up and for your suggestions.

During development it is very convenient, that jte templates are recompiled/hot reloaded whenever stuff in the templates changes, without the need to trigger maven or restart the server.

I also think the limited complexity having to put just one artifact (< 60KB) into the pom.xml is very nice atm :-)

How would you setup the maven project with jte-compiler and jte-rt? I think it would mean to have different maven profiles for development and production, right?

Still I'm wondering if we could support code like this with two dependencies (that's how I use jte on my website at the moment):

´´´
codeResolver = new DirectoryCodeResolver(TEMPLATE_DIRECTORY);
if (environmentPlugin.is(Environment.Debug)) {
    templateEngine = TemplateEngine.create(codeResolver, CLASS_DIRECTORY);
} else {
    templateEngine = TemplateEngine.createPrecompiled(CLASS_DIRECTORY);
}
´´´

**Comment by bendem** [comment]
> How would you setup the maven project with jte-compiler and jte-rt? I think it would mean to have different maven profiles for development and production, right?

The easy way would be two maven modules, not profiles. You'd have a root pom declaring those two modules and would have the compiler module depend on the runtime module.
That way if you need class reloading, you depend on the compiler which pulls in the runtime transitively, and if you only have precompiled classes, you depend on the runtime.

**Comment by casid** [comment]
I thought about users of the library, not the library itself. What I like a lot about the current implementation, is that you simply put the jte dependency in your project and you can right aways start developing (writing tests, firing up a dummy site) and you instantly see the impacts on the template. With separate dependencies the initial project setup would be a lot more complicated, I think.

I just had a look at the size of the rocker runtime dependency and that's 65KB. So I think jte with 60 KB isn't that problematic in terms of size.

I'm very open for suggestions though if the split can be done while leaving the user's project setup / development on the user's project as simple as it is now!

**Comment by bendem** [comment]
My answer is valid for the users of the library as well since they use maven in the exact same way. They either include `jte` (which pulls `jte-runtime` transitively) if they want to hot-compile or `jte-runtime` if they pre-compile.

**Comment by bendem** [comment]
> I just had a look at the size of the rocker runtime dependency and that's 65KB. So I think jte with 60 KB isn't that problematic in terms of size.

This is more about not polluting the classpath with classes that are not meant to be used anyway.
And if modularizing your project (you or a library user), the modules required for `jte-runtime` are much fewer than the full project (which will most likely include the compiler module).

**Comment by casid** [comment]
Got your point. But still wondering how users will do it :-)

Cause in my project I will have a hard dependency on `jte` all time, cause I need to start a server and I want to have hot reloading during development. And I will use the same pom.xml to deploy to production, but with a different setting of the engine.

I'm not sure how I can have a different dependency for production builds.

**Comment by bendem** [comment]
Untested, but I'm guessing something like this would work:

```xml
<properties>
  <jte.artifact>jte</jte.artifact>
  ...
</properties>
...

<dependencies>
  ...
  <dependency>
    <groupId>jte</groupId>
    <artifactId>${jte.artifact}</artifactId>
  </dependency>
</dependencies>

<profiles>
  <profile>
    <id>production</id>
    <properties>
      <jte.artifact>jte-runtime</jte.artifact>
    </properties>
  <profile>
<profiles>
```

**Comment by casid** [comment]
This morning I started a <a href="https://github.com/casid/jte/tree/modularization">modularization branch</a> and split `jte` in `jte-runtime` and `jte`.

It doesn't look terrible so far and I think I already spotted one bug! I think currently the Exception template line number lookup will return nothing when running precompiled only.

Next step will be to add some test it, the current tests only test template execution with the compiler.

Would be cool if you'd take a look at it. Any feedback / suggestions are very appreciated :-)
